rfcs/records.md

```diff
@@ -0,0 +1,196 @@
+
+To create a record, you need to use a record constructor. This is where the draft design has four options, each goes with one of the syntax options.
+
+A1. Creation uses special syntax, `new Record { field = value ... }`. This is unsurprising and easy to implement, but verbose.
+A2. Creation uses existing Lua DSL syntax `Record { field = value, ... }`. This is concise but requires a slightly intricate bytecode design to keep efficient.
+B1. Creation uses special syntax, `new Record(value, ...)`. This is unsurprising and easy to implement, but verbose.
+B2. Creation uses existing Lua call syntax, `Record(value, ...)`. This is concise and reasonably easy to keep efficient.
+
+The big difference between variants A and B is whether you need to spell out field names at construction time. There's precedents for going either way; some
+languages like F#/C#/Kotlin implement record construction as a function call and when you define a record, you essentially define the record constructor. This
+is beautifully concise, but is a bit more difficult to migrate away from tables, and it's easy to mix up the names. Variants A are more verbose and either require
+an extra `new` context-specific keyword, or complex/awkward magic to keep construction efficient.
+
+In variants A, it would make sense to allow omission of any field, in which case it gets replaced with the default of `nil`. A future extension could be made to
+allow specification of default values at record definition time. Type checker would fail to type check record construction if fields that have non-optional types
+have the values omitted.
+
+In variants B, it would probably make sense to require exact number of values to be specified, or follow the usual function call syntax rules.
+
+Note that since records are first class objects, you can export or import a record through a module boundary in the usual way:
+
+```
+local X = require(path).X
+local r = new X(1, 2)
+```
+
+### Type checking records
+
+Records defined via a `record` statement can be used in type annotations as usual. The unification rules say that a table can unify with a record if the fields
+match, which makes records similar to sealed tables from the type checking perspective. (note, this is hand wavy on subtyping rules)
+
+> TODO: How do you export a record type? `export record` would be straightforward but potentially conflicts with future export statements for functions/values.
+
+The methods defined on the record object are type checked as usual, with one exception - the implicit `self` has the type of the record. This is crucial because
+this is the one big issue we can't resolve with metatable-based OOP for tables, but it works for records because `:` is slightly more magical.
+
+When `self` is explicit, the type needs to be specified manually, e.g. these definitions are equivalent:
+
+```
+function Point:sum(): number
```

Comment:
This is off topic but, wouldn't it be nice to have a different symbol to represent return types? 

Multiple colons in a sequence makes the code difficult to read at a first glance. Something like:

```lua
function Point:sum() -> number
    return self.x + self.y
end

function Point.sum(self: Point) -> number
    return self.x + self.y
end
```

might look better. Although, it does constitute a significant change to the syntax.

Reply:
I don't know if this is possible to change now unless you allow the `:` syntax to keep working. That said, Rust uses `->` for return types and I find myself typing `:` accidentally all the time, so if Luau also started using `->`, I'd start mistyping in Luau too xD

Reply:
We used to use `=>` early on in Luau history but switched to `:`. I don't think we'd want to change this to `->`. Note that you are already going to have many colons if you have many parameters, and also annotating the return type is actually rarely necessary since the type checker can infer it - it's annotated here for clarity of the RFC.